flow/multiagent_envs/traffic_light_grid.py

```diff
@@ -0,0 +1,260 @@
+        intersection, edge_number, density traffic light state.
+        - For edges in the network, gives the density and average velocity.
+        - For the self.num_local_lights number of nearest lights (itself
+        included), gives the traffic light information, including the last
+        change time, light direction (i.e. phase), and a currently_yellow flag.
+        """
+        # Normalization factors
+        max_speed = max(
+            self.k.scenario.speed_limit(edge)
+            for edge in self.k.scenario.get_edge_list())
+        grid_array = self.net_params.additional_params["grid_array"]
+        max_dist = max(grid_array["short_length"], grid_array["long_length"],
+                       grid_array["inner_length"])
+
+        # TODO(cathywu) refactor PO_TrafficLightGridEnv with convenience
+        # methods for observations, but remember to flatten for single-agent
+
+        # Observed vehicle information
+        speeds = []
+        dist_to_intersec = []
+        edge_number = []
+        all_observed_ids = []
+        for _, edges in self.scenario.node_mapping:
+            local_speeds = []
+            local_dists_to_intersec = []
+            local_edge_numbers = []
+            for edge in edges:
+                observed_ids = \
+                    self.k_closest_to_intersection(edge, self.num_observed)
+                all_observed_ids.append(observed_ids)
+
+                # check which edges we have so we can always pad in the right
+                # positions
+                local_speeds.extend(
+                    [self.k.vehicle.get_speed(veh_id) / max_speed for veh_id in
+                     observed_ids])
+                local_dists_to_intersec.extend([(self.k.scenario.edge_length(
+                    self.k.vehicle.get_edge(
+                        veh_id)) - self.k.vehicle.get_position(
+                    veh_id)) / max_dist for veh_id in observed_ids])
+                local_edge_numbers.extend([self._convert_edge(
+                    self.k.vehicle.get_edge(veh_id)) / (
+                    self.k.scenario.network.num_edges - 1) for veh_id in
+                                           observed_ids])
+
+                if len(observed_ids) < self.num_observed:
+                    diff = self.num_observed - len(observed_ids)
+                    local_speeds.extend([0] * diff)
+                    local_dists_to_intersec.extend([0] * diff)
+                    local_edge_numbers.extend([0] * diff)
+
+            speeds.append(local_speeds)
+            dist_to_intersec.append(local_dists_to_intersec)
+            edge_number.append(local_edge_numbers)
+
+        # Edge information
+        density = []
+        velocity_avg = []
+        for edge in self.k.scenario.get_edge_list():
+            ids = self.k.vehicle.get_ids_by_edge(edge)
+            if len(ids) > 0:
+                # TODO(cathywu) Why is there a 5 here?
```

Comment:
Same as above. I'd think we could remove this 5?

Reply:
I might keep it for now since presumably this was working for the grid benchmarks.